pkg/metrics/collector/sockstat.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2018 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type sockStatCollector struct {
+	socketInUseDesc    *prometheus.Desc
+	socketTimeWaitDesc *prometheus.Desc
+	socketMemoryDesc   *prometheus.Desc
+}
+
+var pageSize = os.Getpagesize()
+
+func init() {
+	registerCollector("socket", NewSockStatCollector)
+}
+
+// NewSockStatCollector returns a new Collector exposing socket stats.
+func NewSockStatCollector() (Collector, error) {
+	return &sockStatCollector{
+		socketInUseDesc: prometheus.NewDesc(
+			"num_inuse_sockets",
+			"Number of inuse sockets",
+			[]string{"protocol"}, nil,
+		),
+		socketTimeWaitDesc: prometheus.NewDesc(
+			"num_tw_sockets",
+			"Number of sockets in time wait state",
+			nil, nil,
+		),
+		socketMemoryDesc: prometheus.NewDesc(
+			"socket_memory",
+			"Amount of memory used by sockets",
+			nil, nil,
+		),
+	}, nil
+}
+
+type socketStats struct {
+	tcpInUse        uint64 // num of tcp in use sockets
+	udpInUse        uint64 // num of udp in use sockets
+	tcpTimeWait     uint64 // num of tcp timewait sockets
+	memUsedInBlocks uint64 // sum of tcp and udp memory used in blocks
+}
+
+func getSockStats(fileName string) (*socketStats, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseSockStats(file)
+}
+
+func (s *socketStats) merge(o *socketStats) {
+	s.tcpInUse += o.tcpInUse
+	s.udpInUse += o.udpInUse
+	s.tcpTimeWait += o.tcpTimeWait
+	s.memUsedInBlocks += o.memUsedInBlocks
+}
+
+func extractSockStatUint(field string, s string, out interface{}) error {
+	v, err := strconv.ParseUint(s, 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid value %q for field %q in sockstats", s, field)
+	}
+	outValue := out.(*uint64)
+	*outValue = v
+	return nil
+}
+
+type sockStatEntry struct {
+	out     interface{}
+	extract func(string, string, interface{}) error
+}
+
+func extractSockStatFields(fields []string, entries map[string]*sockStatEntry) error {
+	for i := 0; i < len(fields) && i+1 < len(fields); i += 2 {
+		k, v := fields[i], fields[i+1]
+		if e, ok := entries[k]; !ok {
+			continue
+		} else if err := e.extract(k, v, e.out); err != nil {
```

Comment:
you don't need else if if you have a `continue` above

Reply:
Used else because otherwise cannot use the variable e. Changed the code to increase the scope of variable e as the following
e, ok := entries[k]
if !ok {
	continue
} 
if err := e.extract(k, v, e.out); err != nil {